src/backend/metal/src/native.rs

```diff
@@ -226,7 +232,64 @@ pub enum DescriptorSetBinding {
 }
 
 #[derive(Debug)]
-pub enum Memory {
+pub struct Memory {
+    pub(crate) heap: MemoryHeap,
+    pub(crate) allocations: Arc<Mutex<MemoryAllocations>>,
+    pub(crate) mapping: Mutex<Option<MemoryMapping>>,
+}
+
+impl Memory {
+    pub(crate) fn new(heap: MemoryHeap) -> Self {
+        Memory {
+            heap,
+            allocations: Arc::new(Mutex::new(MemoryAllocations {
+                starts: BTreeMap::new(),
+                ends: BTreeMap::new(),
+            })),
+            mapping: Mutex::new(None),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct MemoryMapping {
+    pub(crate) range: Range<u64>,
+    pub(crate) buffer: metal::Buffer,
+    pub(crate) location: NSUInteger,
+    pub(crate) length: NSUInteger,
+}
+
+#[derive(Debug)]
+pub(crate) struct MemoryAllocations {
+    starts: BTreeMap<u64, (u64, metal::Buffer)>,
+    ends: BTreeMap<u64, (u64, metal::Buffer)>,
+}
+
+impl MemoryAllocations {
+    pub(crate) fn find(&self, range: Range<u64>) -> Vec<(Range<u64>, metal::Buffer)> {
+        /// Get all unique buffers that intersects specified range
+        let mut buffers = Vec::new();
+        buffers.extend(self.starts.range(range.clone()).map(|(&start, &(end, ref b))| (start .. end, b.clone())));
+        let range = (Bound::Excluded(range.start), Bound::Included(range.end));
+        buffers.extend(self.ends.range(range).map(|(&end, &(start, ref b))| (start .. end, b.clone())));
+        buffers.sort_unstable_by_key(|&(_, ref b)| b.as_ptr());
+        buffers.dedup_by_key(|&mut (_, ref b)| b.as_ptr());
+        buffers
+    }
+
+    pub(crate) fn insert(&mut self, range: Range<u64>, buffer: metal::Buffer) {
+        self.starts.insert(range.start, (range.end, buffer.clone()));
```

Comment:
Depending on the cost of cloning that buffer an additional `Vec<metal::Buffer>` might be handy and only storing indices in the trees.

Reply:
Cloning cost for buffer should be similar to `Arc`